File: algorithms/[203]移除链表元素/solution.ts
// topics = ["链表"]

export class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 * 迭代
 * time O(n), space O(1), n 为链表长度
 */
function removeElements(head: ListNode | null, val: number): ListNode | null {
  const dummy = new ListNode();
  dummy.next = head;

  let node = dummy;
  while (node.next) {
    if (node.next.val == val) {
      node.next = node.next.next;
    } else {
      node = node.next;
    }
  }

  return dummy.next;
}

File: algorithms/[203]移除链表元素/solution.py
# topics = ["链表", "递归"]

from __future__ import annotations
from typing import Union


class ListNode:
    def __init__(self, x: int = 0):
        self.val = x
        self.next: Union[ListNode, None] = None


class Solution:
    def removeElements(self, head: Union[ListNode, None], val: int) -> Union[ListNode, None]:
        """
        递归
        time O(n), space O(n), n 为链表长度
        """
        if not head:
            return head
        head.next = self.removeElements(head.next, val)
        return head.next if head.val == val else head
